subsys/net/l2/ethernet/dsa/dsa.c

```diff
@@ -0,0 +1,202 @@
+/** @file
+ * @brief DSA related functions
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_dsa, CONFIG_NET_DSA_LOG_LEVEL);
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <net/net_core.h>
+#include <net/ethernet.h>
+#include <net/net_mgmt.h>
+#include <net/dsa.h>
+
+/*
+ * Helper functions
+ */
+struct dsa_context *dsa_get_context(struct net_if *iface)
+{
+	struct ethernet_context *eth_ctx;
+
+	eth_ctx = net_if_l2_data(iface);
+	if (eth_ctx == NULL) {
+		LOG_ERR("Iface %p context not available!", iface);
+		return NULL;
+	}
+
+	return eth_ctx->dsa_ctx;
+}
+
+/*
+ * Store, in the ethernet_context for master interface, the original
+ * eth_tx() function, which will send packet with tag appended.
+ */
+int dsa_register_master_tx(struct net_if *iface, dsa_send_t fn)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+
+	ctx->dsa_send = fn;
+	return 0;
+}
+
+#ifdef CONFIG_NET_L2_ETHERNET
+bool dsa_is_port_master(struct net_if *iface)
+{
+	/* First check if iface points to ETH interface */
+	if (net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET)) {
+		/* Check its capabilities */
+		if (net_eth_get_hw_capabilities(iface) &
+		    ETHERNET_DSA_MASTER_PORT) {
+			return true;
+		}
+	}
+
+	return false;
+}
+#else
+bool dsa_is_port_master(struct net_if *iface)
+{
+	return false;
+}
```

Comment:
This will need to be in header file. If Ethernet is not enabled, then this file should not be compiled at all so the check would not do anything here.

Reply:
The only place where `dsa_is_port_master()` is used outside DSA related code (i.e. protected with `CONFIG_NET_DSA`) is
the packet_socket.c file. However, this file is going to be removed (as suggested in the other review reply) and the `connection.c` is going to be rewritten to avoid usage of this function.

(The declaration of `dsa_is_port_master()` is already placed in `include/net/dsa.h` file)